README.rst

```diff
@@ -415,6 +415,9 @@ Variables you might want to override
   obtained via the ``mkversion.sh`` by default.
 * ``PKGVERSION`` is the version to be used when creating packages. It's
   obtained via the ``VERSION`` variable by default.
+* ``PKGITERATION`` is the iteration of the package to be created.  By default
+  it is set to the distro code (e.g. ``trusty``, ``xenial``).  It will be
+  ignored if ``PKG_DEFAULTS`` is set manually.
```

Comment:
This doesn't sound right, after all is `PKG_DEFAULTS`, not `PKG_OVERRIDES` or something like that. Would it be too hard to change this?

Reply:
Hmmm.  Let's put it this way: I think I'd rather implement this first in a way that's in line with how Makd currently works, and then look separately at the question of whether Makd _should_ work this way.

For what it's worth, I think the existing behaviour is not unreasonable.  `PKG_DEFAULTS` is perhaps misnamed, but it's a catch-all collection of all the flags to be passed to `fpm`.  The only reason to override this is to make a completely custom set of `fpm` flags.

From this point of view, things like `PKGVERSION` AND `PKGITERATION` only make sense in the context of a predictable approach to how the `fpm` flags are generated.

On the other hand, an alternative approach could be to redefine things so that:

* some flags are always expected to be set (version and iteration being two obvious ones, but things like package type being another), and are overridable by individual variables like `PKGVERSION`, `PKGITERATION`, and `PKGTYPE`

  * the `fpm` flags could be omitted if the variable is unset or empty

* define `PKG_DEFAULTS` in terms of a combination of these obligatory flags, and a new `PKG_DEFAULT_FLAGS` variable to contain the fixed flags currently hardcoded into `PKG_DEFAULTS`.  There are actually only two that meet this definition: `-D-f -D-sdir`.

  * the `-D--deb-changelog=` flag is currently hardcoded, but it would be easy to add a variable to set it (note of course that this is a `deb`-specific setting).

In other words, instead of the current:

```mak
# Default fpm flags. By default it builds Debian packages from files, a Debian
# changelog is provided, and a version and iteration (using the Debian version)
# (defined lazily so we can use target variables)
ifndef PKG_DEFAULTS
DISTRO_CODENAME="$(shell lsb_release -rs | grep rolling || lsb_release -cs)"

# Default package iteration
PKGITERATION ?= "$(DISTRO_CODENAME)"

PKG_DEFAULTS = -D-f -D-sdir -D-t$(PKGTYPE) -D--deb-changelog="$O/changelog.Debian" \
        -D--version="$(PKGVERSION)" \
        -D--iteration="$(PKGITERATION)" \
        -d lsb_release="$(DISTRO_CODENAME)"
endif
```

we could have:

```mak
# Default fpm flags. By default it builds Debian packages from files, a Debian
# changelog is provided, and a version and iteration (using the Debian version)
# (defined lazily so we can use target variables)
ifndef PKG_DEFAULTS
DISTRO_CODENAME="$(shell lsb_release -rs | grep rolling || lsb_release -cs)"

# Default package iteration
PKGITERATION ?= "$(DISTRO_CODENAME)"

PKG_DEFAULT_FLAGS ?= -D-f -D-sdir -D--deb-changelog="$O/changelog.Debian"

PKG_DEFAULTS = -D-t$(PKGTYPE) \
        -D--version="$(PKGVERSION)" \
        -D--iteration="$(PKGITERATION)" \
        -d lsb_release="$(DISTRO_CODENAME)" \
        $(PKG_DEFAULT_FLAGS)
endif
```

This would not avoid the basic fact that overriding `PKG_DEFAULTS` would leave all other settings irrelevant, but it would make a better way possible without breaking change.  We could then strip out the ability to override `PKG_DEFAULTS` in the next major release.

Reply:
The TL;DR answer is: we can and should improve how makd works w.r.t. `PKG_DEFAULTS` but I don't think we should tie it to this PR.  Because any real fix involves a breaking change, and this PR's new feature doesn't need to.